### PR TITLE
Multivariate rational function field

### PR DESCRIPTION
**Pull request by wbhart**
I tried to convert our existing rational function field to work with multivariates as well as univariates. Unfortunately `test_rand` fails. This is due to the fact that the compiler can no longer properly infer the type. The type assertions on `data` and `fraction_field` after creating a `Rat` are now unions.

I see three possible solutions:

1) Remove `test_rand` tests for rational functions and put up with the type not being inferred.

2) Add an additional parameter to the `Rat{T}` type (and `RationalFunctionField` type) for univariate vs multivariate function fields.

3) Create a different type for multivariate rational functions (and function fields), e.g. `MRat{T}`.

Opinions?

### CODE REVIEW

**Comment by thofma** [comment]
I would welcome the possibility to use the type to distinguish between univariate and multivariate rational function fields. So I would prefer 2) or 3), leaning towards 2).

**Comment by wbhart** [comment]
If there are no objections today I will change the implementation to 2 and see if it works ok.

File: src/generic/RationalFunctionField.jl [hunk 1]
```diff
@@ -22,10 +22,10 @@ base_ring(a::Rat) = base_ring(parent(a))
 
 parent(a::Rat) = a.parent
 
-data(x::Rat{T}) where T <: FieldElement = x.d::Frac{dense_poly_type(T)}
+data(x::Rat{T}) where T <: FieldElement = x.d::Union{Frac{dense_poly_type(T)}, Frac{mpoly_type(T)}}
```
Comment (tthsqe12) [inline src/generic/RationalFunctionField.jl:25]: Is this the line you are talking about?
I don't like this line. Probably an extra parameter to the type is in order.

**Comment by wbhart** [comment]
The switch of Rat and RationalFunctionField to two parameter types seems to have been successful. This PR can be reviewed now.

File: src/generic/FunctionField.jl [hunk 2]
```diff
@@ -528,7 +528,8 @@ parent_type(::Type{FunctionFieldElem{T}}) where T <: FieldElement = FunctionFiel
 elem_type(::Type{FunctionField{T}}) where T <: FieldElement = FunctionFieldElem{T}
 
 function base_ring(R::FunctionField{T}) where T <: FieldElement
-   return R.base_ring::RationalFunctionField{T}
+   U = dense_poly_type(T)
+   return R.base_ring::RationalFunctionField{T, U}
```
Comment (thofma) [inline src/generic/FunctionField.jl:532]: That means that a function field field is always a finite extension of a univariate rational function field, which is ok. But then all the function field methods below should not allow `MPolyElem`.

**Comment by HechtiDerLachs** [comment]
@wbhart : Does this also involve the total ring of fractions that you were talking about in the other meeting?

Yesterday, @ulthiel explained to me a bit his wish to have function fields of affine varieties in Oscar. And if your variety is reducible, we will, instead, have the total ring of fractions. I was wondering whether/how this can all be brought into one common framework. And in particular, whether we have to exploit technical gimmicks in different special cases to speed up arithmetic. 

I mean: For the fraction field of a polynomial ring we probably want another data type than for fraction fields of `MPolyQuo`s, right? 

I don't know, whether this is really the right place for these discussions. But this is just to let you know: I'm interested and following the discussion, so: keep me posted!

**Comment by wbhart** [comment]
@HechtiDerLachs I'm not working on the total ring of fractions here. It's on my list of things to try and accomplish, but haven't started work on that yet.

File: src/generic/FunctionField.jl (hunk 2 above)
Comment (wbhart) [reply src/generic/FunctionField.jl:532]: I guess so. I've excluded them for now.

**Comment by tthsqe12** [comment]
Im taking a look tmr morning

**Comment by thofma** [comment]
Yes, it seems that we are using the evaluation functions.

**Comment by thofma** [comment]
It seems that the PR here removed the evaluation function in the univariate case.

**Comment by wbhart** [comment]
What are they supposed to do? They were defined for RingElement but it isn't possible to create a RationalFunctionField unless it is over a FieldElement type. I don't see how they *could* be used.

**Comment by wbhart** [comment]
By the univariate case you mean this one:

```
function evaluate(f::Rat{T}, v::U) where {T <: PolyElem, U <: Integer}
```

This was impossible to invoke because the `Rat` type was defined as follows:

```
mutable struct Rat{T <: FieldElement}
```

I assumed these functions were just left over from some previous version of the code that was no longer used.

**Comment by thofma** [comment]
My bad. It is unrelated to the evaluation. It is just that the type now has more parameters, so the dispatch does not work as before. It is easy fo fix upstream once this is merged here.

**Comment by wbhart** [comment]
Ok cool. Thanks for checking.

So I better work on something else then until we have another breaking release.

I think total ring of fractions was the most requested project. @HechtiDerLachs @fieker

**Review by tthsqe12 — APPROVED** [review]
Looks good. Can we agree that the functions in oscar should have writen as follows, to remain compatible with any number of type parameters we might add in AA?
Can we just fix them this way and not call it a breaking change?
```
# Evaluation:
function (val::TropicalSemiringMap{S, T} where {U,
                                                S <: AbstractAlgebra.Generic.RationalFunctionField{U},
                                                T <: AbstractAlgebra.Generic.Rat{U}})(c)
```

**Comment by thofma** [comment]
I doubt that this should have been written like this. The code will look really awkward if we start assuming that any type could obtain more parameters in the future.

No, this function is in the latest release. So changing it in Oscar master does not make it non-breaking.

**Comment by thofma** [comment]
Superseded by #1198

**Pull request closed without merging**